source/docs/subset_load_balancer.md

```diff
@@ -0,0 +1,263 @@
+balancing over any upstream host in the cluster, or returning no host.
+
+Load balancing within a subset is accomplished by constructing one of the existing load balancer
+types with `Upstream::HostSet` that presents a filtered copy of the upstream hosts. All load
+balancer types except the Original DST load balancer may be used for subset load balancing.
+
+### Fallback
+
+The SLB can be configured with one of three fallback policies. If no subset matching the
+`LoadBalancerContext` is found:
+
+1. `NO_ENDPOINT` specifies that `chooseHost` returns `nullptr` and load balancing fails.
+2. `ANY_ENDPOINT` specifies that load balancing occurs over the entire set of upstream hosts.
+3. `DEFAULT_SUBSET` specifies that load balancing occurs over a specific subset of upstream
+   hosts. If the default subset is empty, `chooseHost` returns `nullptr` and load balancing fails.
+
+During construction, if the fallback policy is `ANY_ENDPOINT`, a default subset is constructed
+using the original `Upstream::HostSet`. If the fallback policy is `DEFAULT_SUBSET`, but the
+configuration does not specify any metadata (e.g. all hosts match), the SLB changes the fallback
+policy to `ANY_ENDPOINT`.
+
+### Selecting Subsets
+
+The initial implementation supports selecting subsets by endpoint metadata provided via EDS.
+
+The configuration specifies a list of subset selectors. Each selector is used, in turn, to create
+subsets of hosts. The selectors exist to limit the combinations of endpoint metadata used for
+creating subsets. We precompute the subsets outside the load balancing path to avoid locking.
+
+Currently the only mechanism for specifying a selector is to provide a list of metadata keys:
+
+``` json
+{
+  "subset_selectors": [
+    { "keys": [ "a", "b" ] },
+    { "keys": [ "x" ] }
+  ]
+}
+```
+
+For each selector, the SLB iterates over the hosts and inspects the host's metadata for the
+`"envoy.lb"` filter. If a host's metadata provides values for each key, a subset is created for the
+metadata. For example, given the selectors above, if a host's metadata contains `{a=1, b=2}`, a
+subset is created for `{a=1, b=2}`. Other hosts with `{a=1, b=2}` are also included in the subset.
+A host with metadata like `{a=1, b=2, x=3}` is included in two subsets (`{a=1, b=2}` and
+{`x=3`}). The same keys may appear in multiple selector entries: it is feasible to have both an
+`{a=1, b=2}` subset and an `{a=1}` subset.
+
+On update, the SLB divides the hosts added into the appropriate subset(s) and triggers udpate
+events on the filtered host sets. The SLB also manages the optional "local HostSet" used for
+zone-aware routing.
+
+The CDS configuration for the subset selectors is meant to allow future extension. For example:
+
+1. selecting endpoint metadata keys by a prefix or other string matching algorithm
+2. using a list-typed metadata value to allow a single endpoint to have multiple values for a
```

Comment:
This means we don't want to do recursive matching when comparing metadata values then?

Reply:
Currently we don't.

This is a way to allow a single endpoint to be part of multiple subsets based on a single key. So an endpoint A with k=1 and endpoint B with k=1,2 would both be part of a subset for k=1 (and B would be in a second subset as well). An argument could be made that this is just how it should always work, but I was thinking that the metadata values would be treated more or less opaquely -- we need to be able to hash them and compare them for equality but otherwise it's just a blob.